libdnf/dnf-context.cpp

```diff
@@ -2177,9 +2186,21 @@ dnf_context_setup(DnfContext *context,
         !dnf_context_set_os_release(context, error))
         return FALSE;
 
+    /* identify correct rpmdb file */
+    for (i = 0; rpmdb_path_variants[i] != NULL; i++) {
+        rpmdb_path = g_build_filename(priv->install_root, rpmdb_path_variants[i], NULL);
+        if (g_file_test(rpmdb_path, G_FILE_TEST_EXISTS))
+            break;
+        rpmdb_path = NULL;
+    }
+
+    /* if we can't identify an rpmdb path for any reason, fallback to legacy value */
+    if (rpmdb_path == NULL) {
+        rpmdb_path = g_build_filename(priv->install_root, "var/lib/rpm/Packages", NULL);
+    }
+
     /* setup a file monitor on the rpmdb, if we're operating on the native / */
     if (g_strcmp0(priv->install_root, "/") == 0) {
-        rpmdb_path = g_build_filename(priv->install_root, "var/lib/rpm/Packages", NULL);
```

Comment:
The file path is used to set signal when change of RPMDB take place. 
```
g_signal_connect(priv->monitor_rpmdb, "changed",
                         G_CALLBACK(dnf_context_rpmdb_changed_cb), context);
```
@pmatilai Please is there any option how to set such a callback inside rpm? I am looking how to not need to know where RPM stores data. Deleted code will not produce any future issue.

Reply:
Rpm doesn't support any "changed" callbacks on the rpmdb. What's the use-case of that? This sounds to me like a symptom of something else (locking probably) needing fixing.

Reply:
This functionality is required for PackageKit daemon. Daemon loads data (installed and available packages) and reload them only when something is changed. Like when something is installed using rpm, PackageKit daemon must get signal that RPMDB was changed.

Reply:
Ack, figures.

Maybe it could simply watch for the entire directory instead.

Reply:
Anyway, the existing additions for rpmdb change monitoring do almost nothing for this particular use-case, they only really help the kind of on-startup cache-validity check that libsolv, yum etc do (heck even apt-rpm did that). Opened a ticket for adding change notifications but that's not going to help right now:

https://github.com/rpm-software-management/rpm/issues/1124